### DISCUSSION TITLE
mayaUsdProxyShape node export

### DISCUSSION THREAD

**Discussion by sylvieChen** [Autodesk/maya-usd#1347] [Q&A]
Hello,
I have a question about mayaUsdProxyShape. 
I found when I export a mayaUsdProxyShape node to usd, then the mesh would not export. 
but if I export a AL_usdmaya_ProxyShape node, then the mesh would be a reference in the exported usd file.

for example,
`
    def Xform "Bush"
    {
    }
`

`def Xform "Bush" (
                prepend references = @/xxx/bush.usda@
)
{
               
}
`            

we are now adding those references on the usd file after using the maya command to export USD. 
I want to ask if this is something to add in the next release? or maybe we are doing it wrong?

Thank you,
Sylvie

**Comment by kxl-adsk** [comment]
Just to make sure I understand the workflow. You want to use many proxy shapes to assemble the scene and then export it to USD? 

We have been focusing on enabling this workflow directly in USD. In Maya 2022, using a single proxy shape, you can start populating your prim, references, and layer hierarchies. Is this something you considered trying?

And to answer your export question. I believe all that is missing is a translator to export proxy shapes as references. We were focusing on native workflows within a single proxy and this hasn't come up high in priorities.

**Reply by sylvieChen** [reply]
Thank you for your reply.

do you mean to do all edits in one proxy shape? so in maya, we always only have a single usd proxy shapes? I didn't think about that way before, maybe I stuck on our pipeline thinking. if that is the direction, then can I ask you the proper way to add references using ufe API?

is it something like this?
```
ufeSel = ufeUtils.ufe.GlobalSelection.get()
for ufeSceneItem in ufeSel:
    proxyShapeContextOps = ufe.ContextOps.contextOps(ufeSceneItem)
    contextItems = proxyShapeContextOps.getItems([])
    contextItemStrings = [c.item for c in contextItems]
    proxyShapeContextOps.doOp(['Add Reference...'])```

**Reply by ppt-adsk** [reply]
Hi @sylvieChen, in fact, if you already want to add references to USD files under a USD proxy shape, we already support this through the right click context menu, as you may have seen.

If that's not what you're after, I don't recommend you add USD references to USD prims using the UFE API.  UFE is an API for Maya to talk to the maya-usd plugin, as Maya knows nothing about USD itself; Maya talks to plugins through UFE.

Inside the maya-usd plugin you have full access to the USD API, so you should use that, as we do when implementing the UFE API inside maya-usd.  See an example here, where the right click context ops add reference is implemented:
https://github.com/Autodesk/maya-usd/blob/c737fee5925ae876ee8ff2eadece3799935dbaa3/lib/mayaUsd/ufe/UsdContextOps.cpp#L310

Hope this helps.
